### PR TITLE
Fix alpha bugs relating to Person and Add

### LINKED ISSUE #112: Bug: `phone` field does not account for numbers starting with 0
Issue:
Numbers starting with 0s are accepted which might not be intended

Reproduce:
`add n/John Doe p/00000000 e/johnd@example.com t/johnDoe12 r/Exco Member f/`

### LINKED ISSUE #113: Bug: `name` field does not account for legitimate names that contain special characters
Issue:
The `name` field is too restrictive and does not support legitimate names like 'osama s/o laden' 

Suggestion:
Maybe we should allow special characters and numbers in the name field see [this](https://v1.designsystem.tech.gov.sg/patterns/names/#:~:text=Guidelines,symbols%20and%20numbers)
Also, the 's/o' example is given in 2103 github [here](https://nus-cs2103-ay2425s1.github.io/website/admin/tp-pe.html#feature-flaws-2:~:text=For%20example%2C%20disallowing%20s/o%20in%20a%20person%20name%20because%20/%20is%20used%20as%20a%20command%20delimiter%20can%20cause%20a%20major%20problem%20if%20the%20input%20is%20expected%20to%20match%20the%20legal%20name%20of%20the%20person.)

### LINKED ISSUE #114: potential bug: `email` field may reject a valid email address
Currently only these special characters (+_.-) are allowed for the {local-part} but a technically valid local-part also accepts these special printable characters ```!#$%&'*+-/=?^_`{|}~``` [full doc for valid address](https://en.wikipedia.org/wiki/Email_address#Sub-addressing:~:text=Force%20(IETF).-,Local%2Dpart,-%5Bedit%5D)

while technically valid, many applications out there don't account for these special characters. Particularly, this is the regex that the Singapore government uses for their application. [src](https://www.designsystem.tech.gov.sg/patterns/email-address#:~:text=Email%20Address-,Email%20Address,-Use%20this%20pattern)
`pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$"`

Also this leads to a rabbit hold of valid email addresses because there is unquoted and quoted address. Quoted address become much more complex to validate, although there is an article showing a solution [here](https://www.prosiglieres.com/posts/rfc-compliant-email-address-validation/).

I'm not sure if it is worth to support all these 'valid' email address but I thought it is worth to highlight them.

### LINKED ISSUE #115: suggestion: Limit length of telegram username
Might not be in scope, but the official telegram usernames range from 5-32 characters [src](https://core.telegram.org/method/account.checkUsername#:~:text=Accepted%20characters%3A%20A%2Dz%20(case%2Dinsensitive)%2C%200%2D9%20and%20underscores.%0ALength%3A%205%2D32%20characters.)

### LINKED ISSUE #120: Contact name allow number
Currently I can create contact with name like "123", but normally the name shouldn't contain numbers, maybe should only allow alphabetic characters and "/" when creating contact name.

image

### LINKED ISSUE #125: Bug: `add` command allows names with more than one space between words
For most if not all names, there is only a single space between words. Having more than one space between words in the name input is likely to be a typo on the user's end. It may make searching using names difficult. 

Perhaps can remove the extra spaces in names.

### PR DESCRIPTION
**Pull request by shizy**
Fixes #125, fixes #120, fixes #115, fixes #114, fixes #113, fixes #112

Added some new test cases relating to the Person 

Note that possible further discussion whether we should allow numbers in names might be required

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 1]
```diff
@@ -40,6 +40,8 @@ public class AddCommand extends Command {
 
     public static final String MESSAGE_SUCCESS = "New person added: %1$s";
     public static final String MESSAGE_DUPLICATE_PERSON = "This person already exists in the address book";
+    public static final String MULTIPLE_SPACE_WARNING =
+            "CAUTION: This contact's name has multiple whitespaces, this might have been unintentional!";
```
Comment (shizy) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:44]: Fix for #125

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 2]
```diff
@@ -57,8 +59,12 @@ public CommandResult execute(Model model) throws CommandException {
         if (model.hasPerson(toAdd)) {
             throw new CommandException(MESSAGE_DUPLICATE_PERSON);
         }
-
         model.addPerson(toAdd);
+        if (this.toAdd.doesNameHaveConsecutiveWhitespaces()) {
```
Comment (shizy) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:63]: Fix for #125

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -81,6 +81,9 @@ public static Telegram parseTelegram(String telegram) throws ParseException {
         if (!Telegram.isValidTelegram(trimmedTelegram)) {
             throw new ParseException(Telegram.MESSAGE_CONSTRAINTS);
         }
+        if (!Telegram.isValidTelegramLength(trimmedTelegram)) {
```
Comment (shizy) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:84]: Fix for #115

File: src/main/java/seedu/address/model/person/Telegram.java [hunk 4]
```diff
@@ -13,12 +13,21 @@ public class Telegram {
...
+            "Telegram handle must be between 5 to 32 characters";
+
     /*
      * The telegram handle cannot have any whitespaces or be empty, and must
      * only contain characters that are alphanumeric or '_'
      */
     public static final String VALIDATION_REGEX = "^(?=\\S)[a-zA-Z0-9_]+$";
 
+    /*
+     * The telegram handle must be between 5-32 characters in length
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Telegram.java:26]: Fix for #115

File: src/main/java/seedu/address/model/person/Email.java [hunk 5]
```diff
@@ -9,7 +9,7 @@
  */
 public class Email {
 
-    private static final String SPECIAL_CHARACTERS = "+_.-";
+    private static final String SPECIAL_CHARACTERS = "+_%.-";
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Email.java:12]: Fix for #114

File: src/main/java/seedu/address/model/person/Name.java [hunk 6]
```diff
@@ -8,15 +8,19 @@
  * Guarantees: immutable; is valid as declared in {@link #isValidName(String)}
  */
 public class Name {
-
     public static final String MESSAGE_CONSTRAINTS =
-            "Names should only contain alphanumeric characters and spaces, and it should not be blank";
+            "Names should only contain non-numeric characters and spaces, and it should not be blank";
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Name.java:12]: Fix for #120, #113

File: src/main/java/seedu/address/model/person/Name.java [hunk 7]
```diff
@@ -8,15 +8,19 @@
...
 
-    /*
-     * The first character of the address must not be a whitespace,
+    /**
+     * The first character of the name must not be a whitespace,
      * otherwise " " (a blank string) becomes a valid input.
      */
-    public static final String VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
+    public static final String VALIDATION_REGEX = "^[^\\d\\s][^\\d]*$";
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Name.java:18]: Fix for #120, #113

File: src/main/java/seedu/address/model/person/Name.java [hunk 8]
```diff
@@ -8,15 +8,19 @@
...
+
+    /**
+     * System should check if the name has multiple consecutive whitespaces, and warn the user
+     */
+    private static final String WHITESPACE_REGEX = "^(?!.* {2})[ \\S]*$";
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Name.java:23]: Fix for #125

File: src/main/java/seedu/address/model/person/Phone.java [hunk 9]
```diff
@@ -11,8 +11,9 @@ public class Phone {
 
 
     public static final String MESSAGE_CONSTRAINTS =
-            "Phone numbers should only contain numbers, and it should be 8 digits";
-    public static final String VALIDATION_REGEX = "\\d{8}";
+            "Phone number should be a valid Singapore phone numbers\n"
+            + "i.e. Phone number should be 8 digits, and start with either 3,6,8, or 9";
+    public static final String VALIDATION_REGEX = "^[3689]\\d{7}$";
```
Comment (shizy) [inline src/main/java/seedu/address/model/person/Phone.java:16]: Fix for #112

File: src/main/java/seedu/address/model/person/Telegram.java [hunk 10]
```diff
@@ -13,12 +13,21 @@ public class Telegram {
             "Telegram handle must only have characters that are alphanumeric "
                     + "or ‘_’, and cannot be empty";
 
+    public static final String LENGTH_CONSTRAINTS =
+            "Telegram handle must be between 5 to 32 characters";
```
Comment (weiliann) [inline src/main/java/seedu/address/model/person/Telegram.java:17]: between 5 and 32

File: src/main/java/seedu/address/model/role/Role.java [hunk 11]
```diff
@@ -8,7 +8,6 @@
  * Guarantees: immutable; name is valid as declared in {@link #isValidRoleName(String)}
  */
 public class Role {
-
     public static final String MESSAGE_CONSTRAINTS = "Role length should be between 1 to 20 characters";
```
Comment (weiliann) [inline src/main/java/seedu/address/model/role/Role.java:11]: between 1 and 20

**Review by weiliann — APPROVED** [review]
LGTM! just some nitpick

**Review by superb-sushi — APPROVED** [review]
LGTM

**Pull request merged**